### PR TITLE
[core worker] Refactor CoreWorker member classes

### PR DESCRIPTION
**Pull request by stephanie-wang**
## What do these changes do?

Refactoring `CoreWorker` class to encapsulate functionality a bit more. Removes references to the parent `CoreWorker` from the `CoreWorkerObjectInterface`, `CoreWorkerTaskInterface`, and `CoreWorkerTaskExecutionInterface`.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by stephanie-wang** [comment]
cc @raulchen @jovany-wang @zhijunfu, please let me know if these changes don't align with what you guys are thinking!

**Review by zhijunfu — COMMENTED** [review]
Looks good to me, thanks for the effort. I left a few minor comments.

File: src/ray/core_worker/common.h [hunk 1]
```diff
@@ -13,13 +14,10 @@ namespace ray {
...
 enum class WorkerType { WORKER, DRIVER };
 
-/// Language of Ray tasks and workers.
-enum class WorkerLanguage { PYTHON, JAVA };
-
 /// Information about a remote function.
 struct RayFunction {
   /// Language of the remote function.
-  const WorkerLanguage language;
+  const Language language;
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/core_worker/common.h]: I see, yes, I think having a single `Language` type is cleaner. Sure, should I move `Language` to a new protobuf file now?

File: src/ray/core_worker/store_provider/plasma_store_provider.cc [hunk 2]
```diff
@@ -7,11 +7,19 @@
...
-    RayletClient &raylet_client)
-    : store_client_(store_client),
-      store_client_mutex_(store_client_mutex),
-      raylet_client_(raylet_client) {}
+    const std::string &store_socket, RayletClient &raylet_client)
+    : raylet_client_(raylet_client) {
+  // TODO(zhijunfu): currently RayletClient would crash in its constructor if it cannot
+  // connect to Raylet after a number of retries, this needs to be changed
+  // so that the worker (java/python .etc) can retrieve and handle the error
+  // instead of crashing.
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/core_worker/store_provider/plasma_store_provider.cc]: Sure, thanks!

File: src/ray/core_worker/common.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhijunfu) [reply src/ray/core_worker/common.h]: never mind, I could take care of it in my next PR:)

**Review by zhijunfu — APPROVED** [review]
LGTM. thanks.

**Comment by pcmoritz** [comment]
@stephanie-wang There is a compile error in one of the tests, do you want to fix it?
```
src/ray/core_worker/mock_worker.cc: In constructor 'ray::MockWorker::MockWorker(const string&, const string&)':
src/ray/core_worker/mock_worker.cc:20:37: error: 'WorkerLanguage' has not been declared
       : worker_(WorkerType::WORKER, WorkerLanguage::PYTHON, store_socket, raylet_socket,
```

File: src/ray/core_worker/common.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (raulchen) [reply src/ray/core_worker/common.h]: FYI, I'm working on `common.proto` right now.

**Comment by raulchen** [comment]
I'm actually in favor of having the CoreWorker and the sub interfaces friend each other. 

Because the public methods in these interfaces are supposed to exposed to Java/Python workers and should be limited. And on the other hand, inside these interfaces, we sometimes want to use other interfaces' private methods. E.g., `DirectActorCallTransport` would need the private api in `ObjectInterface` that puts an object directly in local memory. 

The initial reason why we split `CoreWorker` into sub interfaces was mainly to avoid making `CoreWorker` class too large, not because of encapsulation. 

Does this make sense?

**Comment by stephanie-wang** [comment]
> I'm actually in favor of having the CoreWorker and the sub interfaces friend each other.
> 
> Because the public methods in these interfaces are supposed to exposed to Java/Python workers and should be limited. And on the other hand, inside these interfaces, we sometimes want to use other interfaces' private methods. E.g., `DirectActorCallTransport` would need the private api in `ObjectInterface` that puts an object directly in local memory.
> 
> The initial reason why we split `CoreWorker` into sub interfaces was mainly to avoid making `CoreWorker` class too large, not because of encapsulation.
> 
> Does this make sense?

Sorry, I don't understand your example about the `DirectActorCallTransport`. Can you describe specifically what the inheritance/friend relationships would be in that case? That seems like you would want to make `DirectActorCallTransport` a friend of an `ObjectInterface`, which I think would be okay in certain cases for the reasons you mentioned, but it does not explain why `ObjectInterface` needs to be a friend of `CoreWorker`.

In general, I just want to avoid giving classes access to private fields that they are not currently using; for example, before this PR, the `CoreWorkerTaskInterface` had access to the plasma `store_socket_`, even though it never uses it.

**Comment by raulchen** [comment]
@stephanie-wang Initially, I was thinking that all the sub interfaces would intensively use other interfaces' private methods. After thinking more, this doesn't seem true. I think what you said should be better. Only use friend in particular classes that need it. Thanks.

**Review by raulchen — APPROVED** [review]

**Comment by zhijunfu** [comment]
@stephanie-wang there's still a compile error, may you please fix it?  There are a few such places in `core_worker_test.cc` 

```
src/ray/core_worker/core_worker_test.cc: In member function 'virtual void ray::TwoNodeTest_TestObjectInterfaceCrossNodes_Test::TestBody()':
src/ray/core_worker/core_worker_test.cc:370:42: error: 'WorkerLanguage' has not been declared
   CoreWorker worker1(WorkerType::DRIVER, WorkerLanguage::PYTHON,
```

**Comment by stephanie-wang** [comment]
> @stephanie-wang there's still a compile error, may you please fix it? There are a few such places in `core_worker_test.cc`
> 
> ```
> src/ray/core_worker/core_worker_test.cc: In member function 'virtual void ray::TwoNodeTest_TestObjectInterfaceCrossNodes_Test::TestBody()':
> src/ray/core_worker/core_worker_test.cc:370:42: error: 'WorkerLanguage' has not been declared
>    CoreWorker worker1(WorkerType::DRIVER, WorkerLanguage::PYTHON,
> ```

Oops, thank you!

**Comment by kfstorm** [comment]
FYI, I changed type of language to `ray::rpc::language` in #5034.

**Comment by stephanie-wang** [comment]
> FYI, I changed type of language to `ray::rpc::language` in #5034.

Thanks, I changed it to just use `Language` everywhere. Hope this is okay...

**Pull request merged**